compiler/infra/Assert.cpp

```diff
@@ -165,4 +168,144 @@ namespace TR
       va_fatal_assertion(file, line, condition, format, ap);
       va_end(ap);
       }
+
+   static bool containsNode(TR::Node *node, TR::Node *target)
+      {
+      if (node == target)
+         return true;
+
+      for (int i = 0; i < node->getNumChildren(); i++)
+         {
+         if (containsNode(node->getChild(i), target))
+            return true;
+         }
+
+      return false;
+      }
+
+   static void markInChecklist(TR::Node *node, TR_BitVector &nodeChecklist)
+      {
+      if (nodeChecklist.isSet(node->getGlobalIndex()))
+         return;
+
+      nodeChecklist.set(node->getGlobalIndex());
+
+      for (int i = 0; i < node->getNumChildren(); i++)
+         markInChecklist(node->getChild(i), nodeChecklist);
+      }
+
+   void NodeAssertionContext::printContext() const
+      {
+      if (!_node) return;
+
+      static bool printFullContext = feGetEnv("TR_AssertFullContext");
+      TR::Compilation *comp = TR::comp();
+      TR_Debug *debug = comp->findOrCreateDebug();
+
+      fprintf(stderr, "\nNode context:\n\n");
+
+      if (printFullContext)
+         {
+         debug->printIRTrees(TR::IO::Stderr, "Assertion Context", comp->getMethodSymbol());
+         debug->print(TR::IO::Stderr, comp->getMethodSymbol()->getFlowGraph());
+         if (comp->getKnownObjectTable())
+            comp->getKnownObjectTable()->dumpTo(TR::IO::Stderr, comp);
+         }
+      else
+         {
+         fprintf(stderr, "...\n");
+
+         TR_BitVector nodeChecklist;
+         nodeChecklist.init(0, comp->trMemory(), heapAlloc, growable);
```

Comment:
Why not a NodeChecklist from the comp object where we have a pool?

Reply:
This bit vector is being used to tell the `TR_Debug` object which nodes have been seen in previous treetops so that the printout correctly marks any commoned nodes. This is done by the call to `restoreNodeChecklist` before printing, but that method expects a `TR_BitVector` instead of a `TR::NodeChecklist` and AFAICT there's no way to convert between the two.

Reply:
I had been thinking this was for the traversal - the usage now seems fine given what TR_Debug needs.